test/experimental/ProbabilisticGraphicalModels/bayesnet.jl

```diff
@@ -1,1225 +1,1576 @@
+                        p_switch = 0.3
+                        p = p_switch * (1 - prev_val) + p_stay * prev_val
+                        return Bernoulli(p)
+                    end,
+                    false,
+                    :discrete,
+                )
+                loop_vars[var] = (;)  # Empty named tuple
+                all_vars[Symbol("z[$i]")] = 0  # Initialize with value 0
+
+                # Add dependency edge
+                add_edge!(bn, prev_var, var)
+            end
+
+            # Add observable at the end
+            y_var = VarName(:y)
+            last_var = VarName(Symbol("z[$n_chain]"))
+
+            add_stochastic_vertex!(
+                bn,
+                y_var,
+                (env, _) -> begin
+                    z_val = AbstractPPL.get(env, last_var)
+                    mu = z_val * 5.0
+                    return Normal(mu, 1.0)
+                end,
+                true,
+                :continuous,
+            )
+            loop_vars[y_var] = (;)  # Empty named tuple
+
+            # Add dependency edge
+            add_edge!(bn, last_var, y_var)
+
+            # Set observation value for y
+            all_vars[:y] = 4.2
+
+            # Create the evaluation environment
+            eval_env = NamedTuple{Tuple(keys(all_vars))}(values(all_vars))
+
+            # Create new bn with full evaluation environment
+            new_bn = BayesianNetwork(
+                bn.graph,
+                bn.names,
+                bn.names_to_ids,
+                eval_env,  # Full environment with all variables
+                loop_vars,
+                bn.distributions,
+                bn.deterministic_functions,
+                bn.stochastic_ids,
+                bn.deterministic_ids,
+                bn.is_stochastic,
+                bn.is_observed,
+                bn.node_types,
+                Dict{VarName,Int}(),  # empty transformed_var_lengths
+                0,  # zero transformed_param_length
+            )
+
+            return new_bn
+        end
+
+        @testset "Correctness with Deep Discrete Graph" begin
+            # Create a chain of discrete variables
+            n_chain = 5
+            bn = create_chain_bayesian_network(n_chain)
+
+            # No continuous parameters in this example
+            params = Float64[]
+
+            # Run both versions
+            _, logp1 = marginalize_without_memo(bn, params)
+            _, logp2, memo_size = marginalize_with_memo(bn, params)
+
+            # Verify results match
+            @test isapprox(logp1, logp2, rtol=1e-10)
+
+            # Make sure there was at least some memoization
+            @test memo_size > 0
+
+            # For a chain of length 5, we could have at most 2*2^n_chain states
+            # The factor of 2 accounts for the fact that each node can be visited 
+            # once per possible configuration of its ancestors
+            @test memo_size <= 2 * 2^n_chain

```

Comment:
I am fine with this, although it feels like `memo_size` should be deterministic? This also applies to the tricky dependency graph test.